core/src/main/java/io/questdb/cairo/wal/WalPurgeJob.java

```diff
@@ -583,4 +416,194 @@ protected boolean runSerially() {
         }
         return false;
     }
+
+    public interface Deleter {
+        void deleteSegmentDirectory(int walId, int segmentId);
+
+        void deleteWalDirectory(int walId);
+    }
+
+    public static class Logic {
+        private final StringSink debugBuffer = new StringSink();
+        private final Deleter deleter;
+        private final LongList discovered = new LongList();
+        private final IntIntHashMap nextToApply = new IntIntHashMap();
+        private final LongList nextToApplyKeys = new LongList();  // LongList rather than IntList because need .sort().
+        private TableToken tableToken;
+
+        public Logic(Deleter deleter) {
+            this.deleter = deleter;
+        }
+
+        public void accumDebugState() {
+            debugBuffer.clear();
+            debugBuffer.put("table=").put(tableToken.getDirName())
+                    .put(", discovered=[");
+
+            for (int i = 0, n = discovered.size(); i < n; i++) {
+                final long encoded = discovered.getQuick(i);
+                final int walId = decodeWalId(encoded);
+                final int segmentId = decodeSegmentId(encoded);
+                final boolean pendingTasks = decodePendingTasks(encoded);
+                final boolean locked = decodeLocked(encoded);
+                if (isWalDir(segmentId)) {
+                    debugBuffer.put("(wal").put(walId);
```

Comment:
nit: you could log next to apply here instead of a separate loop

```suggestion
                    debugBuffer.put("(wal").put(walId).put(",nextToApply:").put(nextToApply.get(walId));
```